libs/stats/odc/stats/io.py

```diff
@@ -242,5 +284,156 @@ def dump(self, task: Task, ds: Dataset, aux: Optional[Dataset] = None) -> Delaye
         sha1_done = self._write_blob(sha1_digest, sha1_url, ContentType="text/plain")
 
         return self._write_blob(
-            json_data, json_url, ContentType=self._meta_contentype, with_deps=sha1_done,
+            json_data, json_url, ContentType=self._stac_meta_contentype, with_deps=sha1_done,
         )
+
+    def dump_with_eodatasets3(self, task: Task, ds: Dataset, aux: Optional[Dataset] = None, proc: StatsPluginInterface = None) -> Delayed:
+        """
```

Comment:
This is meant to be Lazy->Lazy transform, yet it has non-lazy and expensive to compute components that can fail due to network issues.

Reply:
TBH, I am follow this method: https://github.com/opendatacube/odc-tools/blob/develop/libs/stats/odc/stats/io.py#L218 to generate this method. 

Of course, the original method is a lightweight computing component. May I know my version fails because of heavy workload?

Reply:
only non-lazy parts in that method are

- rendering internal structure to json

the rest is all Dask operations. Including COG generation and hash generation from result of cogs. In your case you read pixel data from Dask arrays via `.values`, this forces data transfer from the worker in the best case and a complete recomputation in the worst case.

Reply:
You need to construct a lazy operation for thumbnail generation. Good way to think about lazy vs non-lazy is this:

Say you have  `Y = should_be_lazy(X)`, this should work even if `X.compute()` throws an error. In your case `Y` will not be returned if `X.compute()` throws error, because `should_be_lazy` is not lazy at all.

Reply:
Do you suggests that I can follow this: https://github.com/opendatacube/odc-tools/blob/6e2037b3fb8c48b152d3ca613147ff60d661850e/libs/stats/odc/stats/io.py#L182 design. Just move the thumbnail_generation codes to a new method like:

> def  _ds_to_thumbnail_cog() -> List[Delayed]

to convert the thumbnail generation to a lazy method?

Reply:
Yes, all "dump" operations should be lazy, except for cheap metadata operations: computing output paths and generating eo3 yamls or stac jsons. Any computation that depends on pixel values of the final output should be delayed properly, including thumbnail generation.

Reply:
I am doing this change now.